compiler/optimizer/OMROptimizer.cpp

```diff
@@ -1553,6 +1552,13 @@ int32_t OMR::Optimizer::performOptimization(const OptimizationStrategy *optimiza
             doThisOptimization = true;
          }
          break;
+      case IfCatchBlocks:
+         {
+         TR::SymbolReferenceTable * symRefTab = comp()->getSymRefTab();
+         if (symRefTab->getSymRef(TR::SymbolReferenceTable::excpSymbol))
```

Comment:
Symbol reference table used to (and maybe still) create some symbols up front (regardless of any other conditions). Are we sure that `excpSymbol` is only created when exception (not OSR, or any) catch blocks are present ? I want to make sure we don't apply the optimization when we don't need to.

Reply:
When the optimization is attempted, there is an additional condition:
```      
      if (!b->getExceptionPredecessors().empty() &&
          !b->isOSRCatchBlock() &&
          !b->isEmptyBlock()) // VP may have removed all trees from the block

Reply:
True, but that is after the optimization is entered I presume. Whereas the flag above is to decide whether to even enter the optimization (i.e. call `perform` I presume). I guess what I'm asking is whether the flag `IfCatchBlocks` is a reliable indicator of whether catch block exists or not.

Reply:
Currently, the flag `IfCatchBlocks` does not distinguish between catch blocks from Java and catch blocks inserted by OSR.
I can modify that by using `TR_ResolvedJ9Method::numberOfExceptionHandlers()` API which looks at some info in the ROMMethod. There are two ways of doing this:
1. In inliner, when a callee gets inlined, ask about the number of exception handlers and set a flag in the comp object. I believe the right method to add this code is `TR_InlinerBase::inlineCallTarget`
2. When the filter is tested, scan all inlined methods and see if any had exception handlers.

Any preference between the two above, or any other alternative suggestion?

Reply:
I prefer the 1, since I believe there must be existing code like that that gets used to set a flag such as `IfNews` (tracking new allocations).

Reply:
I pushed the new version of the code. 
I reused the flag `HasExceptionHandlers` from `ResolvedMethodSymbol::_properties` which was not read at all and added a getter/setter. I don't know the distinction between `ResolvedMethodSymbol::_properties` and `MethodSymbol::_methodFlags` where HasNews resides.
The `HasExceptionHandlers` property will be set by OpenJ9 during ILgen.
Inliner will propagate the `HasExceptionHandlers` property from callee to caller, as it does for HasNews.
I also added a getter from the comp object `OMR::Compilation::hasExceptionHandlers()` that can be used by the optimizer to decide if the CatchBlockProfiler optimization should be attempted.
```
bool OMR::Compilation::hasExceptionHandlers() { return self()->getMethodSymbol()->hasExceptionHandlers(); }
```
I could have used the methdSymbol from the optimizer itself, but for HasNews we seem to prefer the query from the comp object.